### PR TITLE
Add segment generator counters to MSQ reports

### PR DESCRIPTION
**Pull request by adarshsanjeev**
Adds segment generator counters as a new type of counters for the MSQ report. These currently tracks the number of rows processed by the segment generator. These could be used to display a more accurate progress bar on the web console for the segment generation stage.

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java:67]: Since its a long, can this ever by null ?

**Review by cryptoe — APPROVED** [review]
Minor comment. Otherwise looks good to me.

File: docs/multi-stage-query/api.md
Comment (cryptoe) [inline docs/multi-stage-query/api.md:535]: Should the counter should have the key "output" and then output type is "segmentGenerationProgress"
Does that interfere with the current metrics ?

Comment (adarshsanjeev) [reply docs/multi-stage-query/api.md:535]: Yes, I think it would cause a problem since we already use that name. I think ideally we should keep the names unique, to avoid any exceptions, since we cast to the child class after fetching based on channel name. 
If we want "output" in the name to make it clear, we could rename this to "segmentGenerationOutput"?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java]: Thanks, I missed this. Changed.

File: docs/multi-stage-query/api.md
Comment (vogievetsky) [reply docs/multi-stage-query/api.md:535]: @cryptoe I do not understand your suggestion. Could you please provide an example of a JSON so help illustrate your point?

The current structure seems good to me although maybe I am missing something.

**Review by vogievetsky — APPROVED** [review]
I tested out the functionality and the API by building the web console part here: https://github.com/apache/druid/pull/13929 . Everything seems to check out.

**Comment by cryptoe** [comment]
@adarshsanjeev Looks like there is a checkstyle failure.

**Review by vogievetsky — CHANGES_REQUESTED** [review]
Actually I changed my approval, I was trying to run this build on slightly larger datasets and it consistently fails. I ran the same query on a fresh build of master with the same config (both builds started via `./bin/start-druid`) and it worked. I assume something is wrong in this branch.

The ingest query I ran is:

```sql
REPLACE INTO "trips_1" OVERWRITE ALL
WITH "ext" AS (SELECT *
FROM TABLE(
  EXTERN(
    '{"type":"http","uris":["https://static.imply.io/example-data/trips/trips_xaa.csv.gz"]}',
    '{"type":"csv","findColumnsFromHeader":false,"columns":["trip_id","vendor_id","pickup_datetime","dropoff_datetime","store_and_fwd_flag","rate_code_id","pickup_longitude","pickup_latitude","dropoff_longitude","dropoff_latitude","passenger_count","trip_distance","fare_amount","extra","mta_tax","tip_amount","tolls_amount","ehail_fee","improvement_surcharge","total_amount","payment_type","trip_type","pickup","dropoff","cab_type","precipitation","snow_depth","snowfall","max_temperature","min_temperature","average_wind_speed","pickup_nyct2010_gid","pickup_ctlabel","pickup_borocode","pickup_boroname","pickup_ct2010","pickup_boroct2010","pickup_cdeligibil","pickup_ntacode","pickup_ntaname","pickup_puma","dropoff_nyct2010_gid","dropoff_ctlabel","dropoff_borocode","dropoff_boroname","dropoff_ct2010","dropoff_boroct2010","dropoff_cdeligibil","dropoff_ntacode","dropoff_ntaname","dropoff_puma"]}',
    '[{"name":"trip_id","type":"long"},{"name":"vendor_id","type":"long"},{"name":"pickup_datetime","type":"string"},{"name":"dropoff_datetime","type":"string"},{"name":"store_and_fwd_flag","type":"string"},{"name":"rate_code_id","type":"long"},{"name":"pickup_longitude","type":"double"},{"name":"pickup_latitude","type":"double"},{"name":"dropoff_longitude","type":"double"},{"name":"dropoff_latitude","type":"double"},{"name":"passenger_count","type":"long"},{"name":"trip_distance","type":"double"},{"name":"fare_amount","type":"double"},{"name":"extra","type":"double"},{"name":"mta_tax","type":"double"},{"name":"tip_amount","type":"double"},{"name":"tolls_amount","type":"double"},{"name":"ehail_fee","type":"string"},{"name":"improvement_surcharge","type":"string"},{"name":"total_amount","type":"double"},{"name":"payment_type","type":"long"},{"name":"trip_type","type":"string"},{"name":"pickup","type":"string"},{"name":"dropoff","type":"string"},{"name":"cab_type","type":"string"},{"name":"precipitation","type":"double"},{"name":"snow_depth","type":"long"},{"name":"snowfall","type":"long"},{"name":"max_temperature","type":"long"},{"name":"min_temperature","type":"long"},{"name":"average_wind_speed","type":"double"},{"name":"pickup_nyct2010_gid","type":"long"},{"name":"pickup_ctlabel","type":"long"},{"name":"pickup_borocode","type":"long"},{"name":"pickup_boroname","type":"string"},{"name":"pickup_ct2010","type":"long"},{"name":"pickup_boroct2010","type":"long"},{"name":"pickup_cdeligibil","type":"string"},{"name":"pickup_ntacode","type":"string"},{"name":"pickup_ntaname","type":"string"},{"name":"pickup_puma","type":"long"},{"name":"dropoff_nyct2010_gid","type":"long"},{"name":"dropoff_ctlabel","type":"long"},{"name":"dropoff_borocode","type":"long"},{"name":"dropoff_boroname","type":"string"},{"name":"dropoff_ct2010","type":"long"},{"name":"dropoff_boroct2010","type":"long"},{"name":"dropoff_cdeligibil","type":"string"},{"name":"dropoff_ntacode","type":"string"},{"name":"dropoff_ntaname","type":"string"},{"name":"dropoff_puma","type":"long"}]'
  )
))
SELECT
  TIME_PARSE("pickup_datetime") AS "__time",
  "trip_id",
  "vendor_id",
  "dropoff_datetime",
  "store_and_fwd_flag",
  "rate_code_id",
  "pickup_longitude",
  "pickup_latitude",
  "dropoff_longitude",
  "dropoff_latitude",
  "passenger_count",
  "trip_distance",
  "fare_amount",
  "extra",
  "mta_tax",
  "tip_amount",
  "tolls_amount",
  "ehail_fee",
  "improvement_surcharge",
  "total_amount",
  "payment_type",
  "trip_type",
  "pickup",
  "dropoff",
  "cab_type",
  "precipitation",
  "snow_depth",
  "snowfall",
  "max_temperature",
  "min_temperature",
  "average_wind_speed",
  "pickup_nyct2010_gid",
  "pickup_ctlabel",
  "pickup_borocode",
  "pickup_boroname",
  "pickup_ct2010",
  "pickup_boroct2010",
  "pickup_cdeligibil",
  "pickup_ntacode",
  "pickup_ntaname",
  "pickup_puma",
  "dropoff_nyct2010_gid",
  "dropoff_ctlabel",
  "dropoff_borocode",
  "dropoff_boroname",
  "dropoff_ct2010",
  "dropoff_boroct2010",
  "dropoff_cdeligibil",
  "dropoff_ntacode",
  "dropoff_ntaname",
  "dropoff_puma"
FROM "ext"
PARTITIONED BY MONTH
```

**Comment by adarshsanjeev** [comment]
@vogievetsky I wasn't able to reproduce this locally with the query supplied. I have also rebased with master, do you still see any issues in your build now?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java:44]: Can we take a lock once and increment all three metrics ?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/SegmentGeneratorFrameProcessor.java
Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/SegmentGeneratorFrameProcessor.java:216]: Lets increment stuff on each added row since we are fetching from an atomic long and the slowest operation would be the merging step anyway.

File: server/src/main/java/org/apache/druid/segment/realtime/FireDepartmentMetrics.java
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/segment/realtime/FireDepartmentMetrics.java:117]: Can you please document what merge rows mean's ?

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java:788]: Why would the queryable index by empty here. 
Would it cause problems in the indexes.add() method ?

Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java:788]: I think the getter above would return null if the queryable index is closed. I think we gracefully handle this elsewhere, I haven't changed that part of the code. This is just to avoid throwing an NPE here if we do run into this problem, since we are just reporting metrics here.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java]: Done

File: server/src/main/java/org/apache/druid/segment/realtime/FireDepartmentMetrics.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/segment/realtime/FireDepartmentMetrics.java:119]: Added a comment to the variable

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/SegmentGeneratorFrameProcessor.java
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/SegmentGeneratorFrameProcessor.java]: I have made the change. I think this would still end up happening to a small extent, if say some rows are not persisted by the time the last rows are processed.

Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/SegmentGeneratorFrameProcessor.java]: This would be really weird behavior for the end user. 
Should we create a class that extends FireDepartMetric and then update the taskReport as and when relevant calls are made ?

Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/SegmentGeneratorFrameProcessor.java]: Changed!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGeneratorMetricsWrapper.java
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGeneratorMetricsWrapper.java:32]: Lets make this final and private ?

**Review by cryptoe — APPROVED** [review]
There is a minor comment but the latest patch LGTM. 
@vogievetsky waiting for your feedback on the latest patch.

**Comment by vogievetsky** [comment]
Looking

**Review by vogievetsky — APPROVED** [review]
I retried the ingestion and it appears to all work now.

I also saw the new metrics. I updated the ui (in https://github.com/apache/druid/pull/13929) to use `rowsMerged` instead of `rowsProcessed` and the UI feedback is much better. There is a bigger gap between the two.

image

This looks good to me!

**Comment by vogievetsky** [comment]
Side note question: should the UI show just `rowsMerged` or should it show `rowsPersisted` also (At this point I think there is no point showing `rowsProcessed` anywhere in the UI)?

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java
Comment (gianm) [inline server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java:789]: This isn't the right place to increment the number of rows merged; it's happening _before_ merge so it's really the number of rows that are about-to-be-merged. If you want the number of rows actually merged, the hook point should be after `mergeQueryableIndex` is done.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java
Comment (gianm) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/counters/SegmentGenerationProgressCounter.java:44]: `incrementRowsProcessed` (spelling)

**Comment by gianm** [comment]
In addition to the comments I left above, IMO it'd also be good to add `rowsPushed`, so we can capture the zip-and-push-to-deep-storage step as well.

**Comment by vogievetsky** [comment]
I second the `rowsPushed` I think the best UI would have 3 lines for the segmentGenerator step:

```
Input
Rows merged
Rows pushed
```

That would be 💯 

Note that I am imagining rows pushed to increment by the number of rows in the segment when it is done pushing.

**Comment by vogievetsky** [comment]
`rowsPushed` would lag behind `rowsMerged` by however much time it takes to push to deep storage which might take some meaningful time in S3 for larger segments. In any case I think it would look great and be very informative.

**Comment by vogievetsky** [comment]
Btw also I love the change to rowsMerged, it now lags quite a bit behind input so it really shows work being done!

**Comment by adarshsanjeev** [comment]
I think this might be because we push to deep storage right after we merge, so the time where they are different is pretty narrow, especially since using local disk as deep storage might be cheaper than normal (if that is how you're running it). Persist and push steps have two different executors processing it, so we might see more of a difference there.

**Comment by vogievetsky** [comment]
Oh cool, I did not realize we push as we merge! (Yes I am doing local deep storage)

**Review by gianm — APPROVED** [review]
LGTM, although I want to point out that these metrics are not updated for streaming ingest. Someone might want them at some point. If you are seeing this from the future: the way to add them to streaming ingest would be to update StreamAppenderator as well.

Btw, nor is it updated for native batch in OPEN_SEGMENTS and CLOSED_SEGMENTS modes. The default is CLOSED_SEGMENTS, so most people would be using that. The place to add them there would be AppenderatorImpl.

Final note: at some point we should remove AppenderatorImpl. It's a maintenance burden to have three similar versions of the same logic.

**Pull request merged**
